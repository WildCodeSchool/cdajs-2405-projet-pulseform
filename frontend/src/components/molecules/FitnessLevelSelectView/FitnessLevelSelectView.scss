@use "@assets/variables" as *;

.fitness-level-select-view {
  height: 100vh;

  &__container {
    &__form {
      display: flex;
      flex-direction: column;
      align-items: stretch;
      height: 100%;
      width: 16rem;
      gap: 1rem;
      margin: 1.5rem 0 1rem;

      @media screen and (min-width: 500px) {
        height: auto;
      }

      &__container {
        display: flex;
        align-items: stretch;
        justify-content: space-between;

        &__value-name {
          @include small-phone;
          display: flex;
          align-items: center;
          justify-content: center;
          background-color: $tertiary;
          color: $white;
          width: 2.5rem;
          height: 2.5rem;
          border-radius: 10px;
        }
      }

      &__btn {
        width: 100%;
      }
    }
  }
}
