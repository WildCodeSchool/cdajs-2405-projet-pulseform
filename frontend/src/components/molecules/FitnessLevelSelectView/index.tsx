import { useMutation } from "@apollo/client";
import { useState } from "react";
import { useTranslation } from "react-i18next";

import BasicButton from "@components/atoms/BasicButton";
import BodyStepQuestions from "@components/molecules/BodyStepQuestions";
import ExitSignUpStep from "@components/molecules/ExitSignUpStep";

import { useUser } from "@context/UserContext";
import { FitnessLevel, MemberRole } from "@graphql/__generated__/schema";
import { UPDATE_USER_MUTATION } from "@graphql/mutations/user";

import "./FitnessLevelSelectView.scss";
import CompleteLaterButton from "../CompleteLaterButton/CompleteLaterButton";

function FitnessLevelSelectView() {
  const [isExit, setIsExit] = useState(false);
  const [isNextStep, setIsNextStep] = useState(false);
  const [lvl, setLvl] = useState<number | null>(null);

  const { t } = useTranslation();
  const { user } = useUser();

  const [updateUser] = useMutation(UPDATE_USER_MUTATION, {
    refetchQueries: ["Me"],
  });

  const handleExit = () => setIsExit(true);

  const handleLevelSelection = (level: number) => {
    setLvl(level);
  };

  const handleNextStep = async () => {
    if (!user || lvl === null) return;

    const levelMap: Record<number, FitnessLevel> = {
      1: FitnessLevel.Beginner,
      2: FitnessLevel.Intermediate,
      3: FitnessLevel.Advanced,
    };

    const selectedLevel = levelMap[lvl];
    const sanitizedWeights =
      user.weights?.map(({ month, weight }) => ({
        month,
        weight,
      })) ?? [];

    try {
      await updateUser({
        variables: {
          data: {
            id: Number(user.id),
            username: user.username,
            email: user.email,
            password: "",
            description: user.description ?? "",
            image: user.image ?? "",
            role: user.role ?? MemberRole.User,
            created_at: new Date(user.created_at),
            birthday: user.birthday ? new Date(user.birthday) : new Date(),
            height: user.height ?? 170,
            weights: sanitizedWeights,
            gender: user.gender ?? "NOT_WISH_TO_ANSWER",
            level: selectedLevel,
          },
        },
      });

      setIsNextStep(true);
    } catch (err) {
      console.error("❌ Erreur lors de la mise à jour du niveau :", err);
    }
  };

  return (
    <>
      {!isExit && !isNextStep && (
        <section className="fitness-level-select-view">
          <BodyStepQuestions
            questionLabel={t("DESCRIBE_YOUR_LEVEL")}
            ctaExit={handleExit}
          >
            <form
              className="fitness-level-select-view__container__form"
              onSubmit={(e) => {
                e.preventDefault();
                handleNextStep();
              }}
            >
              <BasicButton
                typeButton="white"
                className="fitness-level-select-view__container__form__btn"
                type="button"
                onClick={() => handleLevelSelection(1)}
                hasFocus
              >
                {t("BEGINNER")}
              </BasicButton>
              <BasicButton
                typeButton="white"
                className="fitness-level-select-view__container__form__btn"
                type="button"
                onClick={() => handleLevelSelection(2)}
                hasFocus
              >
                {t("INTERMEDIATE")}
              </BasicButton>
              <BasicButton
                typeButton="white"
                className="fitness-level-select-view__container__form__btn"
                type="button"
                onClick={() => handleLevelSelection(3)}
                hasFocus
              >
                {t("ADVANCED")}
              </BasicButton>
              <BasicButton
                className="fitness-level-select-view__container__form__btn"
                type="submit"
                disabled={lvl === null}
              >
                {t("NEXT")}
              </BasicButton>
              <CompleteLaterButton />
            </form>
          </BodyStepQuestions>
        </section>
      )}
      {isExit && !isNextStep && <ExitSignUpStep />}
      {!isExit && isNextStep && <ExitSignUpStep isCompleted />}
    </>
  );
}

export default FitnessLevelSelectView;
