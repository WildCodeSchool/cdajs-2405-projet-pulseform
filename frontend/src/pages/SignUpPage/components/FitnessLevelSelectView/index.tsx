import { useState } from "react";
import { useTranslation } from "react-i18next";

import BasicButton from "@components/atoms/BasicButton";
import BodyStepQuestions from "@components/molecules/BodyStepQuestions";
import ExitSignUpStep from "../ExitSignUpStep";
import FitnessGoalsView from "../FitnessGoalsView";

import "./FitnessLevelSelectView.scss";

function FitnessLevelSelectView() {
  const [isExit, setIsExit] = useState(false);
  const [isNextStep, setIsNextStep] = useState(false);
  const [lvl, setLvl] = useState(0);

  const { t } = useTranslation();

  const handleExit = () => {
    setIsExit(true);
  };

  const handleNextStep = () => {
    setIsNextStep(true);
  };

  const handleLvl = (lvl: number) => {
    setLvl(lvl);
  };

  return (
    <>
      {!isExit && !isNextStep && (
        <section className="fitness-level-select-view">
          <BodyStepQuestions
            questionLabel={t("DESCRIBE_YOUR_LEVEL")}
            ctaExit={handleExit}
          >
            <form
              className="fitness-level-select-view__container__form"
              action=""
            >
              <BasicButton
                typeButton="white"
                className="fitness-level-select-view__container__form__btn"
                type="button"
                onClick={() => handleLvl(1)}
                hasFocus
              >
                {t("BEGINNER")}
              </BasicButton>
              <BasicButton
                typeButton="white"
                className="fitness-level-select-view__container__form__btn"
                type="button"
                onClick={() => handleLvl(2)}
                hasFocus
              >
                {t("INTERMEDIATE")}
              </BasicButton>
              <BasicButton
                typeButton="white"
                className="fitness-level-select-view__container__form__btn"
                type="button"
                onClick={() => handleLvl(3)}
                hasFocus
              >
                {t("ADVANCED")}
              </BasicButton>
              <BasicButton
                className="fitness-level-select-view__container__form__btn"
                type="submit"
                onClick={handleNextStep}
                disabled={lvl === 0}
              >
                {t("NEXT")}
              </BasicButton>
            </form>
          </BodyStepQuestions>
        </section>
      )}
      {isExit && !isNextStep && <ExitSignUpStep />}
      {!isExit && isNextStep && <FitnessGoalsView />}
    </>
  );
}

export default FitnessLevelSelectView;
