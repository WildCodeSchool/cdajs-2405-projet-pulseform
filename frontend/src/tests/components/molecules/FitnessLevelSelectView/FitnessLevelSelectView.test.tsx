import { MockedProvider } from "@apollo/client/testing";
import FitnessLevelSelectView from "@components/molecules/FitnessLevelSelectView";
import { render } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";

test("renders the FitnessLevelSelectView component", () => {
  render(
    <MockedProvider mocks={[]} addTypename={false}>
      <MemoryRouter>
        <FitnessLevelSelectView />
      </MemoryRouter>
    </MockedProvider>,
  );
});
