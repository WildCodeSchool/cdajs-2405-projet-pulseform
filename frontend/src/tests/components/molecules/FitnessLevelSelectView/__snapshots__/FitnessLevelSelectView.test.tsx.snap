// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`FitnessLevelSelectView > renders the FitnessLevelSelectView component 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <section
        class="fitness-level-select-view"
      >
        <div
          class="body-step-questions"
        >
          <div
            class="body-step-questions__header-d"
          >
            <div
              class="body-step-questions__header-d__logo"
            >
              <div
                class="little-logo little-logo--desktop"
              >
                <img
                  alt="Logo Pulse Form"
                  class="little-logo__img"
                  src="/src/assets/icons/mobile-logo.svg"
                />
                <p
                  class="little-logo__title little-logo__title--desktop"
                >
                  Pulse Form
                </p>
              </div>
            </div>
            <div
              class="body-step-questions__header-d__container"
            >
              <button
                class="body-step-questions__header-d__container__btn"
                type="button"
              >
                TEST_PROGRAM
              </button>
            </div>
          </div>
          <button
            class="body-step-questions__exit"
            type="button"
          >
            <img
              alt="Close"
              src="/src/assets/icons/blue-cross.svg"
            />
          </button>
          <h2
            class="body-step-questions__question"
          >
            DESCRIBE_YOUR_LEVEL
          </h2>
          <img
            alt="blob-girl"
            class="body-step-questions__blob-girl"
            src="/src/assets/images/blob-girl.svg"
          />
          <div
            class="body-step-questions__container"
          >
            <form
              action=""
              class="fitness-level-select-view__container__form"
            >
              <button
                class="basic-button fitness-level-select-view__container__form__btn basic-button--white basic-button--white--focus"
                type="button"
              >
                BEGINNER
              </button>
              <button
                class="basic-button fitness-level-select-view__container__form__btn basic-button--white basic-button--white--focus"
                type="button"
              >
                INTERMEDIATE
              </button>
              <button
                class="basic-button fitness-level-select-view__container__form__btn basic-button--white basic-button--white--focus"
                type="button"
              >
                ADVANCED
              </button>
              <button
                class="basic-button fitness-level-select-view__container__form__btn basic-button--basic"
                disabled=""
                type="submit"
              >
                NEXT
              </button>
            </form>
          </div>
          <div
            class="body-step-questions__container-d"
          >
            <div
              class="body-step-questions__container-d__question"
            >
              <img
                alt="sport"
                class="body-step-questions__container-d__question__img"
                src="/src/assets/images/step-img.svg"
              />
              <div
                class="body-step-questions__container-d__question__container"
              >
                <p
                  class="body-step-questions__container-d__question__container__label"
                >
                  MOTIVATIONAL_LABEL
                </p>
              </div>
            </div>
            <div
              class="body-step-questions__container-d__children"
            >
              <p
                class="body-step-questions__container-d__children__label"
              >
                DESCRIBE_YOUR_LEVEL
              </p>
              <form
                action=""
                class="fitness-level-select-view__container__form"
              >
                <button
                  class="basic-button fitness-level-select-view__container__form__btn basic-button--white basic-button--white--focus"
                  type="button"
                >
                  BEGINNER
                </button>
                <button
                  class="basic-button fitness-level-select-view__container__form__btn basic-button--white basic-button--white--focus"
                  type="button"
                >
                  INTERMEDIATE
                </button>
                <button
                  class="basic-button fitness-level-select-view__container__form__btn basic-button--white basic-button--white--focus"
                  type="button"
                >
                  ADVANCED
                </button>
                <button
                  class="basic-button fitness-level-select-view__container__form__btn basic-button--basic"
                  disabled=""
                  type="submit"
                >
                  NEXT
                </button>
              </form>
            </div>
          </div>
        </div>
      </section>
    </div>
  </body>,
  "container": <div>
    <section
      class="fitness-level-select-view"
    >
      <div
        class="body-step-questions"
      >
        <div
          class="body-step-questions__header-d"
        >
          <div
            class="body-step-questions__header-d__logo"
          >
            <div
              class="little-logo little-logo--desktop"
            >
              <img
                alt="Logo Pulse Form"
                class="little-logo__img"
                src="/src/assets/icons/mobile-logo.svg"
              />
              <p
                class="little-logo__title little-logo__title--desktop"
              >
                Pulse Form
              </p>
            </div>
          </div>
          <div
            class="body-step-questions__header-d__container"
          >
            <button
              class="body-step-questions__header-d__container__btn"
              type="button"
            >
              TEST_PROGRAM
            </button>
          </div>
        </div>
        <button
          class="body-step-questions__exit"
          type="button"
        >
          <img
            alt="Close"
            src="/src/assets/icons/blue-cross.svg"
          />
        </button>
        <h2
          class="body-step-questions__question"
        >
          DESCRIBE_YOUR_LEVEL
        </h2>
        <img
          alt="blob-girl"
          class="body-step-questions__blob-girl"
          src="/src/assets/images/blob-girl.svg"
        />
        <div
          class="body-step-questions__container"
        >
          <form
            action=""
            class="fitness-level-select-view__container__form"
          >
            <button
              class="basic-button fitness-level-select-view__container__form__btn basic-button--white basic-button--white--focus"
              type="button"
            >
              BEGINNER
            </button>
            <button
              class="basic-button fitness-level-select-view__container__form__btn basic-button--white basic-button--white--focus"
              type="button"
            >
              INTERMEDIATE
            </button>
            <button
              class="basic-button fitness-level-select-view__container__form__btn basic-button--white basic-button--white--focus"
              type="button"
            >
              ADVANCED
            </button>
            <button
              class="basic-button fitness-level-select-view__container__form__btn basic-button--basic"
              disabled=""
              type="submit"
            >
              NEXT
            </button>
          </form>
        </div>
        <div
          class="body-step-questions__container-d"
        >
          <div
            class="body-step-questions__container-d__question"
          >
            <img
              alt="sport"
              class="body-step-questions__container-d__question__img"
              src="/src/assets/images/step-img.svg"
            />
            <div
              class="body-step-questions__container-d__question__container"
            >
              <p
                class="body-step-questions__container-d__question__container__label"
              >
                MOTIVATIONAL_LABEL
              </p>
            </div>
          </div>
          <div
            class="body-step-questions__container-d__children"
          >
            <p
              class="body-step-questions__container-d__children__label"
            >
              DESCRIBE_YOUR_LEVEL
            </p>
            <form
              action=""
              class="fitness-level-select-view__container__form"
            >
              <button
                class="basic-button fitness-level-select-view__container__form__btn basic-button--white basic-button--white--focus"
                type="button"
              >
                BEGINNER
              </button>
              <button
                class="basic-button fitness-level-select-view__container__form__btn basic-button--white basic-button--white--focus"
                type="button"
              >
                INTERMEDIATE
              </button>
              <button
                class="basic-button fitness-level-select-view__container__form__btn basic-button--white basic-button--white--focus"
                type="button"
              >
                ADVANCED
              </button>
              <button
                class="basic-button fitness-level-select-view__container__form__btn basic-button--basic"
                disabled=""
                type="submit"
              >
                NEXT
              </button>
            </form>
          </div>
        </div>
      </div>
    </section>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
